ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,265 @@
+
+## Specification
+
+### Storage
+
+The `ValidatorSetManager` needs to store all this data to operate correctly.
+
+**Note:** This is pseudo-code that might not be 100% syntactically correct.
+
+```solidity
+// The ID of the Subnet that uses this contract to manage its validator set
+bytes32 subnetID;
+
+// The current Subnet validator set (list of NodeIDs)
+bytes32[] subnetCurrentValidatorSet;
+
+// Subnet validator
+struct SubnetValidator {
+    // The Warp message ID corresponding to the RegisterSubnetValidatorTx that added the validator.
+    // The message ID is needed to submit subsequent SetSubnetValidatorWeightTx.
+    bytes32 messageID,
+    // The nonce of the latest validator weight update
+    uint64 nonce,
+    // The validator weight
+    uint64 weight,
+    // The timestamp at the validator was added
+    uint64 addTimestamp
+}
+
+// The Subnet current validators
+// NodeID => SubnetValidator
+mapping(bytes32 => SubnetValidator) subnetCurrentValidators;
+
+// The total weight of the current Subnet validator set
+uint64 subnetTotalWeight;
+
+// Subnet validator pending registration
+struct SubnetRegisteringValidator {
+    // The validator weight
+    uint64 weight,
+    // The timestamp after which the signature is invalid
+    uint64 expiry,
+    // The Ed25519 signature over the concatenated bytes of [subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]
+    bytes signature
+}
+
+// The Subnet validators that are pending registration (= waiting for addition notification from the P-Chain)
+// NodeID => registering validator
+mapping(bytes32 => SubnetRegisteringValidator) subnetRegisteringValidators;
+
+// Subnet validation status
+enum SubnetValidationStatus {
+    Registering,
+    Active,
+    Removing,
+    Over
+}
+
+// Subnet validation
+struct SubnetValidation {
+    // The validation status
+    SubnetValidationStatus status,
+    // The list of validation periods
+    // The index is the nonce associated with the weight update.
+    SubnetValidationPeriod[] periods,
+    // The nonce of the current validation period
+    uint64 currentValidationPeriod
+}
+
+// Subnet validation period
+struct SubnetValidationPeriod {
+    // The weight of the validator during the period
+    uint64 weight,
+    // The start time of the validation period
+    uint64 startTime,
+    // The end time of the validation period (only ≠ 0 when the period is over)
+    uint64 endTime,
+    // The uptime of the validator during this validation period (only ≠ 0 once the validation period is over)
+    uint64 uptime
+}
+
+/**
+ * The list of message IDs associated with a validator of the Subnet
+ * Each message ID corresponds to a RegisterSubnetValidatorTx
+ * NodeID => messageID[]
+ */
+mapping(bytes32 => bytes32[]) subnetValidatorMessages;
+// The validation corresponding to each message ID (= RegisterSubnetValidatorTx)
+// messageID => SubnetValidation
+mapping(bytes32 => SubnetValidation) subnetValidations;
```

Comment:
```suggestion
// NodeID => messageID => SubnetValidation
mapping(bytes32 => mapping(bytes32 => SubnetValidation) subnetValidations;
```
Here it might be more relevant to add a layer with the `NodeID` to avoid having to query `subnetValidatorMessages` beforehand if we already have the `messageID`